pruning/manager.go

```diff
@@ -0,0 +1,264 @@
+package pruning
+
+import (
+	"container/list"
+	"encoding/binary"
+	"fmt"
+	"sync"
+
+	"github.com/tendermint/tendermint/libs/log"
+	dbm "github.com/tendermint/tm-db"
+
+	"github.com/cosmos/cosmos-sdk/pruning/types"
+)
+
+type Manager struct {
+	db                     dbm.DB
+	logger                 log.Logger
+	opts                   types.PruningOptions
+	snapshotInterval       uint64
+	pruneHeights           []int64
+	// Although pruneHeights happen in the same goroutine with the normal execution,
+	// we sync access to them to avoid soundness issues in the future if concurrency pattern changes.
+	pruneHeightsMx         sync.Mutex
+	// These are the heights that are multiples of snapshotInterval and kept for state sync snapshots.
+	// The heights are added to this list to be pruned when a snapshot is complete.
+	pruneSnapshotHeights   *list.List
+	// Snapshots are taken in a separate goroutine fromt the regular execution
+	// and can be delivered asynchrounously via HandleHeightSnapshot.
+	// Therefore, we sync access to pruneSnapshotHeights with this mutex. 
+	pruneSnapshotHeightsMx sync.Mutex
+}
+
+const errNegativeHeightsFmt = "failed to get pruned heights: %d"
+
+var (
+	pruneHeightsKey         = []byte("s/pruneheights")
+	pruneSnapshotHeightsKey = []byte("s/pruneSnheights")
+)
+
+func NewManager(db dbm.DB, logger log.Logger) *Manager {
+	return &Manager{
+		db:             db,
+		logger:         logger,
+		opts:           types.NewPruningOptions(types.PruningNothing),
+		pruneHeights:   []int64{},
+		pruneSnapshotHeights:   list.New(),
+	}
+}
+
+// SetOptions sets the pruning strategy on the manager.
+func (m *Manager) SetOptions(opts types.PruningOptions) {
+	m.opts = opts
+}
+
+// GetOptions fetches the pruning strategy from the manager.
+func (m *Manager) GetOptions() types.PruningOptions {
+	return m.opts
+}
+
+// GetFlushAndResetPruningHeights returns all heights to be pruned during the next call to Prune().
+// It also flushes and resets the pruning heights.
+func (m *Manager) GetFlushAndResetPruningHeights() ([]int64, error) {
+	if m.opts.GetPruningStrategy() == types.PruningNothing {
+		return []int64{}, nil
+	}
+	m.pruneHeightsMx.Lock()
+	defer m.pruneHeightsMx.Unlock()
+
+	pruningHeights := m.pruneHeights
+
+	// flush the updates to disk so that it is not lost if crash happens.
+	if err := m.db.SetSync(pruneHeightsKey, int64SliceToBytes(pruningHeights)); err != nil {
+		return nil, err
+	}
+
+	m.pruneHeights = make([]int64, 0, m.opts.Interval)
+
+	return pruningHeights, nil
```

Comment:
That's why I  do this:
```
pruningHeights := m.pruneHeights
...
m.pruneHeights = make([]int64, 0, m.opts.Interval)
```

So that  `m.pruneHeights` gets a new underlying buffer and we can safely return `pruningHeights`

Reply:
That rests on the assumption that `m.pruneHeights` will never be modified in-place and will always be replaced outright. That may be true today, but there's no good way to ensure it's upheld as an invariant over time.

Reply:
Changing this to:
```
pruningHeights := make([]int64, m.opts.Interval)
copy(pruningHeights, m.pruneHeights)
m.pruneHeights = m.pruneHeights[:0]
```

Reply:
Changed to copy as suggested:
```
pruningHeights := make([]int64, len(m.pruneHeights))
copy(pruningHeights, m.pruneHeights)
m.pruneHeights = m.pruneHeights[:0]
```